### ISSUE TITLE
Feature request: Add ability to run custom commands when certain events happen (break starts/finishes/gets skipped/postponed)

### ISSUE THREAD

**Issue by hughdavenport** [hovancik/stretchly#712]
### Prerequisites

- [x] I'm using latest version: v 1.1.99
- [x] I've checked [Known issues](https://github.com/hovancik/stretchly#known-issues)
- [x] I've checked [Open issues](https://github.com/hovancik/stretchly/issues)

### Description

More of a feature request. I would be interested in running a custom command (for my use case, it is a screen locker) when you are on the long break (and/or when you postpone/skip a break, or when the break finishes?). My use case is so that when a long break starts, I can walk away without worrying that my computer is unlocked.

Operating system: Gentoo

### Steps to Reproduce

1. Run Stretchly
2. Wait until a long break starts
3. Walk away from your computer

**Expected behavior:** My computer will run a custom command when the lock break starts, which then locks my screen.

**Actual behavior:** No custom command is run, my screen just locks as it normally would without Stretchy (I have a timer, but would like more fine grained integration)

**Reproduces how often:** 100%

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

I'd be happy contributing the patch if it is something that would be accepted.

**Comment by hovancik** [comment]
Hi there :) 

I wanna create a system where breaks will be "plugins" allowing different kind of things. So right now I don't plan to do any new specific things. I need to find time and create new issue where I will put together thoughts and all things people requested for break behavior.

**Comment by vikrantpogula** [comment]
@hovancik  - not sure about what your contributor policy is but I would be happy to work on this feature since I have a similar requirement! Let me know your thoughts

**Comment by hovancik** [comment]
Hey hi @vikrantpogula, here's some basics: https://github.com/hovancik/stretchly#development

But with this feature request, I wanna work toward something that I described in previous comment. I really need to find time to put all requests together ;/

**Comment by vikrantpogula** [comment]
cool thanks @hovancik  - let me know when you collect some ideas!

**Comment by leikoilja** [comment]
Upvoting this idea. Would really want to run some custom bash commands at certain events(like when the break is finished). Would start by "work started" event to enable  `freedom.io` software using a simple API call to enable distraction off mode on all my devices.
@hovancik, please keep us updated of your availability, perhaps taking it in baby steps and enabling "custom_commands" on specific events for now?

**Comment by jhubcnam** [comment]
Upvoting the idea. In fact, I came here with an opposite idea: "disable the screen-saver/locker during Strechly breaks, so that I can see the remaining time of my break". But OP's idea is probably better/safer (depends if the user is in an alone setting or not, I guess).

I may also help by contributing code, I will try to keep up when a plan will be presented. Thanks for the work!

**Comment by hovancik** [comment]
Hi guys, TBH, I was thinking to that "plugin" system that would allow different kind of breaks but it has not moved anywhere. 

Anyway, how do you think this should work? User would specify command somewhere and then Stretchly runs it? Just thinking about it, not sure how useful would safe it would be.

**Comment by leikoilja** [comment]
Hey, @hovancik. I saw some similar implementation in other tool and found it super useful. How about we would have a separate tab "Commands"(or something like that), where you can create an event and a handle for it. For example, choosing from the event dropdown with options like "break starts", "break finishes", "break skipped" etc. and in the command box you can enter a custom command-line command that will be run when the even happens. So for example, I would create a new entity under "Commands" tab, where I would choose event to be "long break starts" and command to be `echo 'Long break has started'`

**Comment by hovancik** [comment]
Yes, that's the easiest way to implement. Node has `exec`: https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback 

I'm just not sure whether it's good/secure idea to do so :) 

does anyone knows if it were possible to have such a thing as Stretchly emitting some kind of event and other apps can react to it? I would feel much better with this behavior.

**Comment by leikoilja** [comment]
@hovancik, let's hear if somebody else knows a better way, but i see no problem having straight `exec` from Stretchly. If you feel it's insecure from the Stretchly side, perhaps one can have that functionality disabled to begin with and only allowed to be enabled with user consent at their own risk?

**Comment by raphaelfournier** [comment]
@hovancik I like your idea of Stretchly emitting a signal at certain events. A complementary idea would be to modify the command-line interface with a command "pinging" Stretchly, which would send a formatted message about its state in response. A single line, "B-3:45-5:00" could mean we're on a break, with 3:45 elapsed out of 5:00. A Json could be even better, but perhaps overkill.
Something like `stretchly --state`.

It would be very secure. However, I have not looked at your code, I don't know how this is currently handled and if it is easily feasible.

Then, for most purposes (I guess), one could write a Bash wrapper (and/or use Cron), to perform actions based on Stretchly's state.

**Comment by hovancik** [comment]
Showing something in command line is impossible right now, I think, because there is some magic with passing parameters from second-instanceto the first and output happens in there. 

Looks like we should use [sockets](https://nodejs.org/api/net.html#net_class_net_socket): 
- Stretchly acts as server and sends messages on specific events, ie: break started, ended, posponed, breaks pauses, breaks resets etc..
- Client (user) can then react to those messages by connecting to socket and listening for messages
   - we could write some basic examples for Mac/Windows/Linux users
- later we can add ability for client to send message and server can response (as mentioned `state`)

I don't have experience with node and sockets, but if it were stable with electron, looks like good solution to me. I would prefer not to run any command that is written by user, because it then opens a chance to write a virus that simply edits Stretchly's config and runs weird stuff. Obviously virus could also react to Stretchly's messages from socket, but at least it's not Stretchly who runs the ommand :)

**Comment by hovancik** [comment]
Stretchly now has commnd line interface, so adding `state` command is possible now.

**Comment by hovancik** [comment]
Possible status messages: 

When in Break

`break mini {remaining ms}`
`break long {remaining ms}`

When paused

`paused {remaining ms}` - manually
`paused indefinitely` - manually but indef.
`paused dnd` - because of DnD 
`paused idle` - because of idle time
`paused exclusion`- because of some appExclusion

Before break

`next mini notification {remaining ms} break {remaining ms}` - when before next mini break and notification enabled
`next mini break {remaining ms}` - when before next mini break and notification enabled
`next long notification {remaining ms} break {remaining ms}` - when before next long break and notification enabled
`next long break {remaining ms}` - when before next long break and notification enabled

What do you think?

**Comment by whispy** [comment]
@hovancik these look great! I'd love to be able to watch the currently Stretchly state and then run certain commands/tasks based on the result. For example, I want to have a notification sent to my smartwatch when Stretchly exits a mini or long break (so that I know I can go back to my computer). This will solve the issue!

Do you have any ETA on when this will be in the app? Any blockers?

**Comment by hovancik** [comment]
Hi, sorry for superlate reply. I got too many messages for Stretchly. 

The only blocker is time, which I have almost zero
